### PR TITLE
fixes #600 - add a configuration option to mergeNodes that allows to …combine relationships of the same type and direction

### LINKED ISSUE #600: add a configuration option to mergeNodes that allows to combine relationships of the same type and direction
So in case of duplicate relationships you don't end up with several of them at the end.

The default would be to not change them
And then there would be the option to combine them by type/direction if the properties are the same
And combine them if the properties differ and then apply the refactoring-config property merge thing

### PR DESCRIPTION
**Pull request by AngeloBusato**
@jexp for merging rels, procedure mergeNodes read the config parameter *mergeRels: true/false*.

If relationships have same start and end nodes are merged into one and properties combined.

ex. with this data set (we want to merge nodes :Person):
apoc refactor mergenodes createdatasetfirstexample

we have 
apoc refactor mergenodes resultfirstexample

------

if relationships have different end or start nodes (related to direction) are maintained both and properties combine into all relationships

ex. with this data set  (we want to merge nodes :Person):

apoc refactor mergenodes createdatasetsecondexample

we have: 
apoc refactor mergenodes resultsecondexampledata

apoc refactor mergenodes resultsecondexample

### CODE REVIEW

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 1]
```diff
@@ -335,12 +336,22 @@ public void categorize(
...
+    private Node mergeNodes(Node source, Node target, boolean delete, Map<String, Object> config) {
         try {
             Map<String, Object> properties = source.getAllProperties();
             copyRelationships(source, copyLabels(source, target), delete);
+            RefactorConfig conf = new RefactorConfig(config);
             if (delete) source.delete();
             PropertiesManager.mergeProperties(properties, target, conf);
+
+            //check mergeRels parameter
+            Object value = config.get("mergeRels") == null ? false : config.get("mergeRels");
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: this is really weird, that info should come cleanly out of RefactorConfig please move it there and add a dedicated method

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 2]
```diff
@@ -335,12 +336,22 @@ public void categorize(
...
+            if(value.equals(true)) {
+                Map<String,Object> map = new HashMap<String, Object>();
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: singletonMap

File: src/main/java/apoc/refactor/GraphRefactoring.java (hunk 1 above)
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: getOrDefault -> in refactorConfig

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 3]
```diff
@@ -396,4 +407,23 @@ private Relationship copyRelationship(Relationship rel, Node source, Node target
         Relationship newrel = startNode.createRelationshipTo(endNode, rel.getType());
         return copyProperties(rel, newrel);
     }
+
+
+
+    public void mergeRelsWithSameTypeAndDirectionInMergeNodes (Node node, RefactorConfig config, Direction dir){
+        List<Relationship> rels = new ArrayList<>();
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: this code is really weird, please fix it
use foreach
iterate over the iterator

File: src/main/java/apoc/refactor/GraphRefactoring.java [hunk 4]
```diff
@@ -396,4 +407,23 @@ private Relationship copyRelationship(Relationship rel, Node source, Node target
...
+        node.getRelationships(dir).iterator().forEachRemaining(rels::add);
+        for (int i = 0; i < rels.size()-1; i++) {
+            for (int j = i + 1; j < rels.size(); j++) {
+                if (rels.get(i).isType(rels.get(j).getType())){
```
Comment (jexp) [inline src/main/java/apoc/refactor/GraphRefactoring.java]: don't call rels.get() repeatedly, use local variables.
they should have the same start/end-node given that you iterate starting from that one

File: src/test/java/apoc/refactor/GraphRefactoringTest.java [hunk 5]
```diff
@@ -475,4 +475,65 @@ public void testMergeRelsEagerAggregationCombineArrayDifferentValuesTypeProperti
...
+    @Test
+    public void testMergeNodesAndMergeSameRelationship() {
+        db.execute("create (a1:ALabel {name:'a1'})-[:HAS_REL {p:'r1'}]->(b1:BLabel {name:'b1'})," +
+                "          (a2:ALabel {name:'a2'})-[:HAS_REL{p:'r2'}]->(b1)," +
+                "           (a3:ALabel {name:'a3'})<-[:HAS_REL{p:'r3'}]-(b1)," +
+                "           (a4:ALabel {name:'a4'})-[:HAS_REL{p:'r4'}]->(b4:BLabel {name:'b4'})");
+        Map<String,Object> params = null;
+
+        testCall(db, "MATCH (a1:ALabel {name:'a1'}), (a2:ALabel {name:'a2'}), (a3:ALabel {name:'a3'}), (a4:ALabel {name:'a4'}) " +
+                        "     WITH head(collect([a1,a2,a3,a4])) as nodes CALL apoc.refactor.mergeNodes(nodes,{properties:\"combine\",mergeRels:true}) yield node return node",
```
Comment (jexp) [inline src/test/java/apoc/refactor/GraphRefactoringTest.java]: use single quotes

File: src/test/java/apoc/refactor/GraphRefactoringTest.java [hunk 6]
```diff
@@ -475,4 +475,65 @@ public void testMergeRelsEagerAggregationCombineArrayDifferentValuesTypeProperti
...
                 });
     }
+
+    @Test
+    public void testMergeNodesAndMergeSameRelationship() {
+        db.execute("create (a1:ALabel {name:'a1'})-[:HAS_REL {p:'r1'}]->(b1:BLabel {name:'b1'})," +
+                "          (a2:ALabel {name:'a2'})-[:HAS_REL{p:'r2'}]->(b1)," +
+                "           (a3:ALabel {name:'a3'})<-[:HAS_REL{p:'r3'}]-(b1)," +
+                "           (a4:ALabel {name:'a4'})-[:HAS_REL{p:'r4'}]->(b4:BLabel {name:'b4'})");
+        Map<String,Object> params = null;
```
Comment (jexp) [inline src/test/java/apoc/refactor/GraphRefactoringTest.java]: what is this form?

File: src/test/java/apoc/refactor/GraphRefactoringTest.java [hunk 7]
```diff
@@ -475,4 +475,65 @@ public void testMergeRelsEagerAggregationCombineArrayDifferentValuesTypeProperti
...
+                "     (n7:City {name:'London'})," +
+                "     (n8:City {name:'Liverpool'})," +
+                "     (n1)-[:WORKS_FOR{since:2015}]->(n3)," +
+                "     (n2)-[:WORKS_FOR{since:2018}]->(n4)," +
+                "     (n3)-[:HAS_HQ{since:2004}]->(n7)," +
+                "     (n4)-[:HAS_HQ{since:2007}]->(n8)," +
+                "     (n1)-[:DRIVE{since:2017}]->(n5)," +
+                "     (n2)-[:HAS{since:2013}]->(n6)");
+
+        testCall(db, "MATCH (a1:Person{name:'John'}), (a2:Person {name:'Tom'})" +
```
Comment (jexp) [inline src/test/java/apoc/refactor/GraphRefactoringTest.java]: this statement is werid.

File: src/test/java/apoc/refactor/GraphRefactoringTest.java [hunk 8]
```diff
@@ -475,4 +475,65 @@ public void testMergeRelsEagerAggregationCombineArrayDifferentValuesTypeProperti
...
+                        "     WITH head(collect([a1,a2])) as nodes CALL apoc.refactor.mergeNodes(nodes,{properties:\"overwrite\",mergeRels:true}) yield node MATCH (n)-[r:WORKS_FOR]->(c:Company{name:'Company1'}) MATCH (n1)-[r1:WORKS_FOR]->(c1:Company{name:'Company2'}) return node, n, r ,c,n1,r1,c1 ",
+                row -> {
+                    assertTrue(row.get("node") != null);
+                    assertTrue(row.get("node") instanceof Node);
+                    Node resultingNode = (Node)(row.get("node"));
```
Comment (jexp) [inline src/test/java/apoc/refactor/GraphRefactoringTest.java]: just cast it

File: src/test/java/apoc/refactor/GraphRefactoringTest.java [hunk 9]
```diff
@@ -475,4 +475,65 @@ public void testMergeRelsEagerAggregationCombineArrayDifferentValuesTypeProperti
...
+                    Relationship r = (Relationship)(row.get("r"));
+                    Node n1 = (Node)(row.get("n1"));
+                    Node c1 = (Node)(row.get("c1"));
+                    Relationship r1 = (Relationship)(row.get("r1"));
+                    assertEquals("Tom", resultingNode.getProperty("name"));
+                    assertEquals(0, resultingNode.getDegree(Direction.INCOMING));
+                    assertEquals(4,resultingNode.getDegree(Direction.OUTGOING));
+                    assertEquals(1,c.getDegree(Direction.INCOMING));
+                    assertEquals(true, r.isType(RelationshipType.withName("WORKS_FOR")));
+                    assertEquals(Arrays.asList( 2015L , 2018L).toArray(), new ArrayBackedList(r.getProperty("since")).toArray());
```
Comment (jexp) [inline src/test/java/apoc/refactor/GraphRefactoringTest.java]: just two lists or Sets

**Comment by AngeloBusato** [comment]
@jexp pr updated

**Comment by jexp** [comment]
Thanks a lot

**Pull request merged**